### ISSUE TITLE
FR: Support returning incomplete results after an interrupt

### ISSUE THREAD

**Issue by krlmlr** [r-lib/gh#98]
Scenario: I'm running a query and didn't realize how many results are returned. The progress bar (#26) indicates that it's going to be too long, but I'd like to look at partial results. I'm pressing Escape (or Ctrl + C) but the intermediate results are gone.

As discussed in #86, we could store intermediate results in a private environment and expose with a function, e.g. `gh_partial()` . Catching interrupts is a bad idea.

**Comment by gaborcsardi** [comment]
I am not convinced that this is a good idea. It seems hard to implement it in a way that ensures that the incomplete storage belongs to the last interrupted `gh()` call.

**Comment by gaborcsardi** [comment]
Catching interrupts could actually work, except that curl is buggy: https://github.com/jeroen/curl/issues/216

**Comment by gaborcsardi** [comment]
Sorry, what I meant above is that returning the incomplete results is a great idea, but storing them in a private place and making sure that they belong to the last request seems hard. 

E.g. what if you get an interrupt before anything was downloaded? Then the previous incomplete results are there. What if you call `gh()` from async code (eg. the async package)? Any kind of concurrency makes this hard.

Maybe we can fix the interrupt issue in curl, and then catch the interrupt. That would be pretty cool, because you could even continue the program, from the browser, with `.tryResumeInterrupt()`.

**Comment by jeroen** [comment]
It may be easier to implement this with the async curl api, like so:

```r
get_data <- function(){
  url <- 'https://nghttp2.org/httpbin/drip?duration=10&numbytes=500'
  pool <- curl::new_pool()
  buf <- rawConnection(raw(0), "r+")
  on.exit({
    out <- rawConnectionValue(buf)
    close(buf)
    return(out)
  })
  curl::curl_fetch_multi(url, pool = pool, data = function(x, ...){
    writeBin(x, buf)
  })
  curl::multi_run(pool = pool)
}

# Interrupt this after a few sec:
get_data()
```

**Comment by gaborcsardi** [comment]
Yeah, but why do the callbacks of the multi api support interrupt handlers and the easy api callback don't?

**Comment by jeroen** [comment]
I just don't really know a good way to do implement this in C.

**Comment by gaborcsardi** [comment]
You could use an approach that is in the cleancall package, and just call `R_CheckUserInterrupt()` normally.

**Comment by gaborcsardi** [comment]
We could just return the result, with a warning, instead of storing it somewhere?

**Comment by hadley** [comment]
I forgot where, but I know we've used this pattern elsewhere (i.e. on `interrupt`, warn and return the current values).

**Comment by gaborcsardi** [comment]
My current thinking is that we could attach the partial results to the “last error” object. Which would be a  “last interrupt” object in this case.

**Comment by shikokuchuo** [comment]
In the past I've used the `return()` called from `on.exit()` pattern. In the context of a function that doesn't return and is expected to be user-interrupted (as it just runs in a loop): https://github.com/shikokuchuo/ichimoku/blob/38325ed6051786709a81c286bce4fb39538a46bf/R/oanda.R#L340

**Comment by gaborcsardi** [comment]
@shikokuchuo ~~what are the pros and cons of this?~~

So, that basically means ignoring the interrupt and returning partial results. Pros:
- easy to implement
- maybe semantics are ok for top level calls.

Cons:
- cannot use it in downstream code, because it makes semantics weird.

**Comment by hadley** [comment]
Here's what we do in httr2: https://github.com/r-lib/httr2/blob/43e0d280d2a58154c42cddb8e6a40374b39ab3b6/R/req-perform-sequential.R#L74-L101

`check_repeated_interrupt()` is a relatively recent innovation that I added because if you're calling this in a loop it's possible to get stuck in a state that it's hard to escape (i.e. it relies on you hitting Ctrl + C when in the outer loop not inside the function): https://github.com/r-lib/httr2/blob/43e0d280d2a58154c42cddb8e6a40374b39ab3b6/R/utils.R#L338-L344.

**Comment by gaborcsardi** [comment]
This is what I have in #237: 

```r
> gh::gh("/users", per_page = 10, .limit = Inf)
50 items, page 5 | 1.4s
^C

! `gh()` interrupted after fetching 50 records.
ℹ Partial results are available in `rlang::last_error()$gh_result`.

> length(rlang::last_error()$gh_result)
[1] 50
> rlang::last_error()$gh_result[[1]][1:3]
$login
[1] "mojombo"

$id
[1] 1

$node_id
[1] "MDQ6VXNlcjE=“
```

If the interrupt is caught, then the partial results are in the `gh_result` slot of the interrupt condition. I think I like this more than the alternatives.
